standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java

```diff
@@ -1408,6 +1410,43 @@ private boolean isUpdateOrDelete(Statement stmt, String conflictSQLSuffix) throw
     }
   }
 
+  public long getLatestTxnInConflict(long txnid) throws MetaException {
+    Connection dbConn = null;
+    Statement stmt = null;
+
+    try {
+      dbConn = getDbConn(Connection.TRANSACTION_READ_COMMITTED);
+      stmt = dbConn.createStatement();
+
+      String writeConflictQuery = "SELECT MAX(\"COMMITTED\".\"WS_TXNID\")" +
```

Comment:
This is probably a big enough change, but I was wondering if you could return all the conflicting txnIds and writeIds. In that case if the max conflicting txnId is less than the current one, you could skip two other call to the HMS for validTxnList and validWriteIdList and just remove the exceptions from the list.

Reply:
In most of the cases it will just decrease the performance. It could only be useful if there is a conflict with already registered in snapshot txn.